Fixing C2v, adding stereographic plots for every point group, and making SYmmetry match ITOC standards

argerlt opened the pull request:
#### Description of the change
This started as an update to the [plot_symmetry_operations](https://orix.readthedocs.io/en/stable/examples/stereographic_projection/plot_symmetry_operations.html#sphx-glr-examples-stereographic-projection-plot-symmetry-operations-py) example, which has a few errors and also does not match the symbol conventions in the International ables of Crystallography (ITOC).

I decided to procedurally generate plots for all the point groups (most the hard work was already done, thank you @hakonanes for that), and in the process noticed `orix.quaternion.symmetry.C2v` was inconsistently defined compared to the other groups as well as ITOC. the common convention is to place the main cyclic rotation group around the z-axis.

In hunting that down, I also noticed some of the symmetry group names also did not match the ITOC names, and some other very minor inconsistencies in naming conventions, so I cleaned them up along with some other functions. I also added some descriptions to existing functions to help with clairity.

orix.quaternion.symmetry should be much more closely aligned with ITOC once this is pushed. Additionally, the following picture shows the auto-generatred stereographic plots for every point group

![image](https://github.com/user-attachments/assets/73f905b9-a5c8-401e-9792-8aa3ca1ac1f7)


#### Progress of the PR
- [x] [Docstrings for all functions](https://numpydoc.readthedocs.io/en/latest/example.html)
- [x] Unit tests with pytest for all lines
- [x] Clean code style by [running black via pre-commit](https://orix.readthedocs.io/en/latest/dev/code_style.html)

#### Minimal example of the bug fix or new feature

see `orix/examples/stereographic_projection/plot_symmetry_operations.py` and the plot above. 

#### For reviewers
<!-- Don't remove the checklist below. -->
- [ ] The PR title is short, concise, and will make sense 1 year later.
- [ ] New functions are imported in corresponding `__init__.py`.
- [ ] New features, API changes, and deprecations are mentioned in the unreleased
      section in `CHANGELOG.rst`.
- [ ] Contributor(s) are listed correctly in `__credits__` in `orix/__init__.py` and in
      `.zenodo.json`.

argerlt committed:
Correct C2v to match ITOC

see chapter 12, as wel as figure 10.3.2 of International Tables of Crystallography, Volume A, for details on this discrepency, but The Cyclic rotation axis should be the z axis, not the x axis, in the standard form.

argerlt committed:
Correct C2v to match ITOC

see chapter 12, as wel as figure 10.3.2 of International Tables of Crystallography, Volume A, for details on this discrepency, but The Cyclic rotation axis should be the z axis, not the x axis, in the standard form.

argerlt committed:
matching symmetry lists and names to ITOC

argerlt committed:
matching symmetry lists and names to ITOC

argerlt committed:
update _get_laue_group_name to be proceedural instead of a name-based lookup table

argerlt committed:
update _get_laue_group_name to be proceedural instead of a name-based lookup table

argerlt committed:
expand symmetry marker and add example

argerlt committed:
expand symmetry marker and add example

argerlt committed:
formatting

argerlt committed:
formatting

argerlt committed:
make tables sphinx-compatable

argerlt committed:
make tables sphinx-compatable

argerlt committed:
formatting

argerlt committed:
formatting

hakonanes reviewed (changes requested):
Very good job. I have some suggestions which I think should improve maintainability and the user experience.

argerlt commented on orix/plot/stereographic_plot.py:
```diff
@@ -630,34 +633,35 @@ def stereographic_grid(
+            if inner != "fill":
+                bg_kwargs = deepcopy(kwargs)
+                if "color" in bg_kwargs:
+                    bg_kwargs.pop("color")
```
fixed

argerlt commented on orix/plot/stereographic_plot.py:
```diff
@@ -968,3 +972,124 @@ def _order_in_hemisphere(polar: np.ndarray, pole: int) -> Union[np.ndarray, None
+    crystallographic point groups.
+
+    Intended to be used indirectly in
+    :func:`~orix.plot.StereographicPlot.symmetry_marker`.
```
agreed

argerlt commented on orix/plot/stereographic_plot.py:
```diff
@@ -968,3 +972,124 @@ def _order_in_hemisphere(polar: np.ndarray, pole: int) -> Union[np.ndarray, None
+        folds=2,
+        inner="fill",
+    ):
+        assert np.isin(folds, [1, 2, 3, 4, 6])
```
can you elaborate? not sure what you mean by this

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -136,10 +138,12 @@ def euler_fundamental_region(self) -> tuple:
             "211": (360,  90, 360),  # Monoclinic
             "121": (360,  90, 360),
             "112": (360, 180, 180),
+              "2": (360, 180, 180),
```
Yes, but a bunch of the checks in Symmetry don't actually check for symmetry operators, they just check `Symmetry.name`. The ITOC name for C2 is `2`, but functions throughout orix, and also likely in people's own code, would break if i removed '112', so I just added both to the lookup table for redundancy.

This is also why there is both `symmetry.C2` and `symmetry.C2z`

I think the right thing to eventually do (though not as part of this PR) would be to remove all name-based lookup tables form functions in `Symmetry`, but that is going to be a major project.

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -670,51 +718,395 @@ def plot(
-# fmt: on
-_proper_groups = [C1, C2, C2x, C2y, C2z, D2, C4, D4, C3, D3x, D3y, D3, C6, D6, T, O]
+
+def get_point_groups(subset: str = "unique"):
```
I'm having troubles seeing your vision, but I'm always down to get rid of a big ugly lookup function.

Can you lay out a super basic skeleton of the class you are suggesting? I'm guessing it's something like this but maybe you have something else in mind:
```
class PointGroups()
    def __init__(name: str |List(str))
    -give PG names as strings, returns list of PGs
    def get (name: str |List(str))
    - classmethod that runs get_point_group_from_space_group and returns PGs
    def subset(str)
    - classmethod that returns a subset list 
```

hakonanes replied:
My suggestion was simply to change @argerlt's function into a class. The class has a constructor that takes one or more subsets (all, Laue, crystal system, etc.), provides all the relevant properties like HM, Schoenflies, Laue class, crystal system etc., can list available point groups in table (string representation), and extract a point group from the list.

E.g., if we want to plot the IPF color key for all Laue classes, we create a table of those point groups only and iterate over them via their HM symbol.

argerlt committed:
add suggestions from #563 review

argerlt committed:
add suggestions from #563 review

argerlt commented:
I think I've tentatively resolved everything except the following two:
https://github.com/pyxem/orix/pull/563/files#r2182622890
https://github.com/pyxem/orix/pull/563/files#r2182562898

These are both going to require more more thought a feedback.

argerlt commented on examples/stereographic_projection/plot_symmetry_operations.py:
```diff
@@ -3,51 +3,130 @@
+
+
+# Iterate through the 32 Point groups
+for i, pg in enumerate(point_groups):
```
Okay, I thought about this for a long time. You are right this should, somehow, be functionalized. However, I don't want a single hardcoded "add symmetry markers" function because:

1) This example follows ITOC, which is marginally different from "Structure of Materials" (see `mmm` for example) and also neither of those methods hold up perfectly when applied to off-axis or non-standard plot orientations (do you plot all the inversion markers? or only some? what if all the rotation axes are all in-plane with xy?).
2) I made some choices with regard to coloring that others might not like. It would be nice to allow users to control those.
3) I think my method might actually break a bit for systems where the z-axis is not pointing out of the page, specifically in how the rotation markers are orientated.

I think I'm going to try breaking this into two parts:

`Symmetry.plottable_symmetry_elements()`  to run the algorithm to find the elements to plot

`StereographicPlot.all_symmetry_markers(sym)` which will run the output of the above function in a `for` loop and have some additional choices for the user to change, but also explicitly show how to run the equivalent with more control using the existing `StereographicPlot.symmetry_markers(sym)`

argerlt replied:
AH! I misunderstood you the first time! I thought you were suggesting adding an option for automatic markers  to `StereographicPlot.scatter`, `StereographicPlot.plot`, etc, which I wasn't crazy about. What you actually suggested was replacing the function `Symmetry.plot`. That's a great plan, I did not even realize that was a function! Sorry, should have read more carefully.

Okay, yup, I'm 100% on board now. I'll fix that up ASAP

> Do you have an example in which off-axis or non-standard plot orientations are used?

Not anymore, I was vaguely worried if someone wanted a custom `StereographicPlot` at an off-axis view, it could potentially mess up how the markers get rotated, as well as rules for where you do and don't add inversion dots to symmetry markers, but for the finite set of possible crystallographic `Symmetry` plots, i think it's fine.

> We can add coloring options for the various symmetry operation types as parameters (rotation, mirror, inversion).

Agreed

> Currently we only support the c-axis out of plane, so these plots are all in line with the current supported crystal axes alignments.

Also agreed

hakonanes replied:
Yes, something like that sounds good. I'd even make the stereographic plot method private! We can instead give more flexibility to users when they ask for it.

argerlt committed:
Merge branch 'develop' into UpdatingSymmetryWithExamples

argerlt committed:
moving plot_symmetry_operations algo to Symmetry function

argerlt committed:
moving plot_symmetry_operations algo to Symmetry function

argerlt committed:
formatting

argerlt committed:
formatting

argerlt commented on orix/plot/stereographic_plot.py:
```diff
@@ -968,3 +972,124 @@ def _order_in_hemisphere(polar: np.ndarray, pole: int) -> Union[np.ndarray, None
+        v: Union[Vector3d, np.ndarray, list, tuple],
+        size: int = 1,
+        folds=2,
+        inner="fill",
```
how about `modifier: Literal[None,"none","rotation","rotoinversion","inversion""] = "none")` then? first three map to None, second two map to rotoinversion and inversion.

argerlt commented:
@hakonanes and/or @CSSFrancis , recent realization: `Symmetry.plot` doesn't do what I think we think it does. 

I assumed it would make a stereographic plot of symmetry elements. What it actually does is make a z-axis pole figure. 

This doesn't make sense to have as a function of the `Symmetry` class, so I think it should be removed.

Do you think it is fine to just overwrite it with this new `Symmetry.plot`? Or, should I depreciate `Symmetry.plot` and call this new function I am writing `Symmetry.plot_elements`?

I'd rather do it the first way, but there is a (very unlikely) world where that might break something for someone downstream
 
Example:
```
from orix.quaternion import Orientation
from orix.quaternion.symmetry import Oh
import matplotlib.pyplot as plt
from orix.vector import Vector3d

# generate some random orientations
oris = Orientation.random(10, Oh)

# run symmetry.plot
Oh.plot(oris)

# Compare that image to a  <100> pole figure
fig, ax = plt.subplots(subplot_kw={'projection':"stereographic"})
ax.scatter(oris.symmetry.outer(oris*Vector3d.zvector()))
```

![image](https://github.com/user-attachments/assets/ebe28561-388e-46dc-8d74-d11921339559)

argerlt committed:
create PointGroups class, improve Symmetry.plot, and fix example

argerlt committed:
create PointGroups class, improve Symmetry.plot, and fix example

argerlt committed:
formatting

argerlt committed:
formatting

hakonanes commented:
The plot is meant to show equivalent directions in a crystal, given one direction. Just like the points in the plots from McHenry and De Graef: https://github.com/pyxem/orix/issues/456#issuecomment-1699489533.

I think the option to add:
1. symmetry operations (new default)
2. equivalent directions
3. both

would be best.

argerlt committed:
finishing symmetry.plot and adding to PointGroups

argerlt committed:
finishing symmetry.plot and adding to PointGroups

argerlt committed:
typos

argerlt committed:
typos

argerlt committed:
formatting

argerlt committed:
formatting

argerlt committed:
more formatting

argerlt committed:
more formatting

hakonanes reviewed (changes requested):
This PR is large (!), with lots of good work.

In this first review, I've only reviewed the new Symmetry methods and the point group look-up class.

After we've addressed my suggestions, I'll continue reviewing the symmetry marker class and example and tests.

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -546,175 +807,687 @@ def plot(
+                + f". name = '{name}' is not a valid value."
+            )
+
+    def from_space_group(
```
Ah, I left out `self` intentionally because this was intended to be called as :

```
sym = PointGroup.from_space_group(111)
```

but what I really should have done was include `self` and add the @classmethod decorator so it can be called both from a class and from from an instance of this class.

Fixed. this also looks more in the style of the rest of orix

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -56,6 +58,7 @@ class Symmetry(Rotation):
     """
 
     name = ""
+    _s_name = ""
```
What if instead we make the function `symm.schoenflies` return None when `symm._schoenflies=""`?

I would prefer keeping the internal variable set by `symmetry` as a string, same as `name`, simply because it makes print operations easier to write, especially if you want to string some `.split` or`.join` operations together. The way I wrote `PointGroups.__repr__` is sort of an extreme example of this. You COULD absolutely do a similar operation if the default was `None`, but when i tried it with `PointGroups.__repr__`, I felt it looked much too convoluted. 

I would also say _s_name (an admittedly terrible variable name that I WILL change, good call) should only be changed by the `symmetry` module, same as `Symmetry.name`, so I think searching for `symm._s_name==""` is just as effective as `symm._s_name is not None`.

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -463,80 +587,217 @@ def fundamental_zone(self) -> Vector3d:
-        onto the upper hemisphere.
+        asymetric_vector: Vector3d | None = None,
+        show_name: bool = True,
+        plt_axis: plt.Axes | None = None,
```
> This is something we should support throughout orix, adding to an existing axis or figure. I'll see if I can get to it.

Yes! I would also like this, and have thought about doing it bit by bit. This could also be an easy first PR for someone interested in working on orix as well, since there is already an example and identifying the functions to change is fairly straightforward.

argerlt replied:
Okay , in practice, turns out `axis` is bad because we have a lot of axes in our plotting functions that refer to Vector3d objects. Trying out `ax`, if we don't like it I can change it.

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -463,80 +587,217 @@ def fundamental_zone(self) -> Vector3d:
+        to modify an existing plot, like so:
+
+        >>> import matplotlib.pyplot as plt
+        >>> pg_Oh = PointGroups.get('m-3m')
```
THAT is a cool feature I did not know about.

argerlt committed:
Documentation improvements

argerlt committed:
Merge branch 'UpdatingSymmetryWithExamples' of https://github.com/argerlt/orix into UpdatingSymmetryWithExamples

argerlt commented on orix/quaternion/symmetry.py:
```diff
@@ -463,80 +587,217 @@ def fundamental_zone(self) -> Vector3d:
+        mirror_width: float = 2.0,
+        asymetric_vector_dict: dict = {},
+        asymmetric_vector_size: float = 10.0,
+        itoc_style=True,
```
Fact checking this was a rough existential spiral for me. I've been using ITOC for literally years, and nothing with that name has existed since the 1930's. Looks like my thesis has 40-odd new typos in it.

Fixed (this, not my thesis)

argerlt committed:
Addressing feedback on #563

argerlt commented:
Okay, I believe I have addressed all review comments and everything is covered and passing. 

~~However, this has gotten rather large and hectic, and there are a lot of PR's right now, I'm going to wait until #555 #558, and #575 are merged before looking at this again.~~

Edit: the relevant parts of the above issues have been merged, now waiting on #647

argerlt committed:
fix broken reference to old class feature

argerlt committed:
fix D3h name and improve explination on plot_symmetry_operations.py

hakonanes commented:
No problem leaving it for a few days! I'm not sure you'd pollute our git history with this PR... Given the limited resources we have for this project, I'm fine not thinking too much about that.

argerlt committed:
Merge remote-tracking branch 'upstream/develop' into UpdatingSymmetryWithExamples

argerlt committed:
fixing errors from 0.14 merge

argerlt committed:
more merge fixes, add example, address review comments

argerlt committed:
passing tests.

argerlt commented:
> I was hoping to start this today, 

Well, that was a lie on my part.

This has been sitting around for too long, I'm just going to fix it here. "refactor and split" was a nice dream, but a year in, and I'm ready to move on.

Few outstanding issues:

- Given some time to look back, I think  `PointGroup.__init__`, `PointGroup.get`, and `PointGroup.get_set` are a bit clunky as-is

- `PointGroups('groups')[0].name` should give the name(s), but doesn't. there are some other symmetry values like this that probably need fixing.

- Language, formatting, etc.

It is far enough along though for a re-review.

argerlt commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
